webhookURLValidator.go

```diff
@@ -0,0 +1,209 @@
+			if err = f(parsedFailureURL); err != nil {
+				return fmt.Errorf("%w: %v", errInvalidFailureURL, err)
+			}
+		}
+		return nil
+	}
+}
+
+// GoodAlternativeURLs parses the given webhook's Config.AlternativeURLs
+// and returns as soon as the URL is considered invalid. It returns nil if the URL is
+// valid.
+func GoodAlternativeURLs(vs []ValidURLFunc) ValidatorFunc {
+	vs = filterNil(vs)
+	return func(w Webhook) error {
+		for _, u := range w.Config.AlternativeURLs {
+			if u == "" {
+				return errInvalidAlternativeURL
+			}
+			parsedAlternativeURL, err := url.ParseRequestURI(u)
+			if err != nil {
+				return fmt.Errorf("%w: %v", errInvalidAlternativeURL, err)
+			}
+			for _, f := range vs {
+				err = f(parsedAlternativeURL)
+				if err != nil {
+					return fmt.Errorf("%w: %v", errInvalidAlternativeURL, err)
+				}
+			}
+		}
+		return nil
+	}
+}
+
+// HTTPSOnlyEndpoints creates a ValidURLFunc that considers a URL valid if the scheme
+// of the address is https.
+func HTTPSOnlyEndpoints() ValidURLFunc {
+	return func(u *url.URL) error {
+		if u.Scheme != "https" {
+			return errURLIsNotHTTPS
+		}
+		return nil
+	}
+}
+
+// RejectHosts creates a ValidURLFunc that checks the URL and ensures the
+// host does not contain any strings in the list of invalid hosts. It returns an error
+// if the host does include an invalid host name.
+func RejectHosts(invalidHosts []string) ValidURLFunc {
+	ih := []string{}
+	for _, v := range invalidHosts {
+		if v != "" {
+			ih = append(ih, v)
+		}
+	}
+	return func(u *url.URL) error {
+		for _, v := range ih {
+			if strings.Contains(u.Host, v) {
+				return errInvalidHost
+			}
+		}
+		return nil
+	}
+}
+
+// RejectALLIPs creates a ValidURLFunc that checks if the URL is an IP and returns an error
+// if it is.
+func RejectAllIPs() ValidURLFunc {
+	return func(u *url.URL) error {
+		host := u.Hostname()
+		ip := net.ParseIP(host)
+		if ip != nil {
+			return errIPGivenAsHost
+		}
+		return nil
+	}
+}
+
+// RejectLoopback creates a ValidURLFunc that returns an error if the given URL is
+// a loopback address.
+func RejectLoopback() ValidURLFunc {
+	return func(u *url.URL) error {
+		host := u.Hostname()
+		ip := net.ParseIP(host)
+		if ip != nil && ip.IsLoopback() {
+			return errLoopbackGivenAsHost
+		}
+		ips, err := net.LookupIP(host)
+		if err != nil {
+			return err
+		}
+		for _, i := range ips {
+			if i.IsLoopback() {
+				return errLoopbackGivenAsHost
+			}
+		}
+		return nil
+	}
+}
+
+// InvalidSubnets checks if the given URL is in any subnets we are blocking and returns
+// an error if it is. SpecialIPs will return nil if the URL is not in the subnet.
+func InvalidSubnets(i []string) (ValidURLFunc, error) {
+	invalidSubnets := []*net.IPNet{}
+	for _, sp := range i {
+		_, n, err := net.ParseCIDR(sp)
+		if err != nil {
+			return nil, errInvalidSubnet
+		}
+		invalidSubnets = append(invalidSubnets, n)
+	}
+	return func(u *url.URL) error {
+		host := u.Hostname()
+		ips, err := net.LookupIP(host)
```

Comment:
does this work for both cases: (1) the hostname is an IP (i.e. 127.0.0.1) and (2) hostname is a domain name  (i.e. localhost)

Reply:
fortunately, LookupIP will work for both hostnames that are IPs and non-IPs.